src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AttachedCollections/IVsHierarchyItemExtensions.cs

```diff
@@ -26,36 +23,54 @@ internal static class IVsHierarchyItemExtensions
         /// <returns><see langword="true"/> if the target was found, otherwise <see langword="false"/>.</returns>
         public static bool TryFindTarget(this IVsHierarchyItem item, [NotNullWhen(returnValue: true)] out string? target)
         {
-            s_targetFlagsRegex ??= new Regex(@"^(?=.*\b" + nameof(Flags.TargetNode) + @"\b)(?=.*\$TFM:(?<target>[^ ]+)\b).*$", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.CultureInvariant);
-
             for (IVsHierarchyItem? parent = item; parent is not null; parent = parent.Parent)
             {
-                if (parent.TryGetFlagsString(out string? flagsString))
+                if (parent.TryGetFlags(out ProjectTreeFlags flags) && flags.Contains(Flags.TargetNode))
                 {
-                    Match match = s_targetFlagsRegex.Match(flagsString);
-                    if (match.Success)
+                    // Found an ancestor target node.
+                    const string prefix = "$TFM:";
+                    string? flag = flags.FirstOrDefault(f => f.StartsWith(prefix));
+
+                    if (flag is not null)
                     {
-                        target = match.Groups["target"].Value;
+                        target = flag.Substring(prefix.Length);
                         return true;
                     }
+
+                    // Target node didn't have a TFM flag for some reason. This is unexpected, but there's not much
+                    // we can do about it. Don't check any more ancestors, as there should only be one relevant target node.
+                    break;
                 }
             }
 
             target = null;
             return false;
         }
 
-        public static bool TryGetFlagsString(this IVsHierarchyItem item, [NotNullWhen(returnValue: true)] out string? flagsString)
+        public static bool TryGetFlags(this IVsHierarchyItem item, out ProjectTreeFlags flags)
         {
-            IVsHierarchyItemIdentity identity = item.HierarchyIdentity;
+            // Browse objects are created lazily, so access the one on the project root in order to obtain the unconfigured project.
+            IVsHierarchyItem root = item;
+            while (!root.HierarchyIdentity.IsRoot)
```

Comment:
i guess we don't need this loop to get IVsHierarchy.  It can be accessed directly from item.HiearchyIdentity.Hierarchy/NestedHierarchy.

Reply:
This is trying to get the project-level hierarchy item, so that we always query the browse object for the project, rather than for each dependency. I can't see a way to access the project-level root without walking nodes, but then the `IVsHierarchyItemIdentity` and related interfaces aren't documented. Do you know a way to do this without looping?